### PR TITLE
Address issue #169

### PR DESCRIPTION
**Pull request by lkwagner**
This PR does two main things: 
1) Now vmc, line_minimization, and dmc all take `client` and `npartition` arguments, which removes some of the pain in running in parallel.

2) The data output of vmc and dmc is now consistent with the HDF output. All functions return a dictionary of numpy arrays, with the first index the step number.

### CODE REVIEW

File: examples/he.py [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
     acc = pyqmc.gradient_generator(mol, wf, to_opt)
     pyqmc.line_minimization(wf, configs, acc, hdf_file="he_opt.hdf5", verbose=True)
+    #pyqmc.vmc(wf, configs)
```
Comment (willwheelera) [inline examples/he.py:16]: what's this for?

File: pyqmc/dmc.py [hunk 2]
```diff
@@ -264,7 +275,7 @@ def rundmc(
     Args:
       wf: A Wave function-like class. recompute(), gradient(), and updateinternals() are used, as well as anything (such as laplacian() ) used by accumulators
 
-      configs: (nconfig, nelec, 3) - initial coordinates to start calculation.
+      configs: (nconfig, nelec, 3) - initial coordinates to start calculation. Will default to
```
Comment (willwheelera) [inline pyqmc/dmc.py:278]: unfinished comment? Will default to...

File: pyqmc/dmc.py [hunk 3]
```diff
@@ -292,68 +303,83 @@ def rundmc(
...
-        df_, configs, weights = propagate(
-            wf,
-            configs,
-            weights,
-            tstep,
-            branchcut_start * esigma,
-            branchcut_stop * esigma,
-            eref=eref,
-            nsteps=branchtime,
-            stepoffset=branchtime * step + stepoffset,
```
Comment (willwheelera) [inline pyqmc/dmc.py:343]: so now "step" in the hdf record is the average of all the steps between branching? the input kwarg nsteps is still npropagate * branchtime, not sure if that's confusing. but we probably don't need each individual step anyway, and the dmc trace will be smoother this way.

File: pyqmc/dmc.py [hunk 4]
```diff
@@ -292,68 +303,83 @@ def rundmc(
...
+        df_["esigma"] = esigma
+        df_["tstep"] = tstep
+        df_['weight_std'] = np.std(weights)
+        df_['nsteps'] = branchtime
+
+        dmc_file(hdf_file, df_, {}, configs, weights)
         # print(df_)
         df.append(df_)
-        eref = df_[ekey[0] + ekey[1]].values[-1].real - feedback * np.log(
+        eref = df_[ekey[0] + ekey[1]] - feedback * np.log(
```
Comment (willwheelera) [inline pyqmc/dmc.py:375]: does eref need to be real?

File: pyqmc/linemin.py [hunk 5]
```diff
@@ -178,14 +180,15 @@ def gradient_energy_function(x, coords):
...
-        en_err = np.std(df["pgradtotal"]) / np.sqrt(len(df))
+        df, coords = vmc(wf, coords, accumulators={"pgrad": pgrad_acc}, client=client, npartitions=npartitions, **vmcoptions)
+        #df = pd.DataFrame(data)[warmup:]
+        en = np.mean(df["pgradtotal"], axis=0)
+        en_err = np.std(df["pgradtotal"], axis=0) / np.sqrt(df['pgradtotal'].shape[0])
         dpH = np.mean(df["pgraddpH"], axis=0)
         dp = np.mean(df["pgraddppsi"], axis=0)
         dpdp = np.mean(df["pgraddpidpj"], axis=0)
         grad = 2 * np.real(dpH - en * dp)
+        print(df["pgraddppsi"].shape)
```
Comment (willwheelera) [inline pyqmc/linemin.py:191]: we probably don't want to keep this print statement

File: pyqmc/linemin.py [hunk 6]
```diff
@@ -178,14 +180,15 @@ def gradient_energy_function(x, coords):
         newparms = pgrad_acc.transform.deserialize(x)
         for k in newparms:
             wf.parameters[k] = newparms[k]
-        data, coords = vmc(wf, coords, accumulators={"pgrad": pgrad_acc}, **vmcoptions)
-        df = pd.DataFrame(data)[warmup:]
-        en = np.mean(df["pgradtotal"])
-        en_err = np.std(df["pgradtotal"]) / np.sqrt(len(df))
+        df, coords = vmc(wf, coords, accumulators={"pgrad": pgrad_acc}, client=client, npartitions=npartitions, **vmcoptions)
+        #df = pd.DataFrame(data)[warmup:]
```
Comment (willwheelera) [inline pyqmc/linemin.py:184]: remove the commented line? or should we check if it's necessary to throw out some warmup steps? seems like if the parameters aren't changing much, the warmup isn't necessary

File: pyqmc/dmc.py [hunk 7]
```diff
@@ -292,68 +303,83 @@ def rundmc(
...
+        with h5py.File(hdf_file, "r") as hdf:
+            stepoffset = hdf["step"][-1] + 1
+            configs.load_hdf(hdf)
+            weights = np.array(hdf["weights"])
+            eref = hdf['eref'][-1]
+            esigma=hdf['esigma'][-1]
+            if verbose:
+                print("Restarted calculation")
+    else:
+        warmup = 2
```
Comment (willwheelera) [inline pyqmc/dmc.py:317]: was this meant to be used two lines below, in en = df[ekey[0]+ekey[1]][2:]? right now it's never used

File: examples/he.py (hunk 1 above)
Comment (lkwagner) [reply examples/he.py:16]: Nothing, removed.

File: pyqmc/dmc.py (hunk 3 above)
Comment (lkwagner) [reply pyqmc/dmc.py:343]: yes, that might be confusing. I think this is the right structure, but the names might need to change.

**Pull request merged**